File: 백준/Gold/1744. 수 묶기/수 묶기.py
import sys
input = sys.stdin.readline
from queue import PriorityQueue

N = int(input())
Plusq = PriorityQueue()
Minusq = PriorityQueue()
one = 0
zero = 0

result = 0

for i in range(N):
    request = int(input())

    if request == 0:
        zero += 1
    elif request == 1:
        one += 1
    elif request > 1:
        Plusq.put(request * -1)
    elif request < 0:
        Minusq.put(request)

pairs = [0] * 2

while Plusq.qsize() > 1:
    pairs[0] = Plusq.get() * -1
    pairs[1] = Plusq.get() * -1

    result += (pairs[0] * pairs[1])

if Plusq.empty() == False:
    result += Plusq.get() * -1

while Minusq.qsize() > 1:
    pairs[0] = Minusq.get()
    pairs[1] = Minusq.get()

    result += (pairs[0] * pairs[1])

if Minusq.empty() == False:
    if zero == 0:
        result += Minusq.get()

result += one
print(result)

File: 백준/Gold/1744. 수 묶기/수 묶기.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.Collections;
import java.util.PriorityQueue;

public class Main {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));

        int N = Integer.parseInt(br.readLine());
        PriorityQueue<Integer> plus = new PriorityQueue<>(Collections.reverseOrder());
        PriorityQueue<Integer> minus = new PriorityQueue<>();
        int zero = 0;
        int one = 0;

        for (int i = 0; i < N; i++) {
            int input = Integer.parseInt(br.readLine());

            if (input == 0)
                zero++;

            else if (input == 1)
                one++;

            else if (input < 0) {
                minus.add(input);
            }
            else {
                plus.add(input);
            }
        }

        int answer = 0;
        while (plus.size() > 1) {
            int first = plus.poll();
            int second = plus.poll();

            answer += first * second;
        }

        if (!plus.isEmpty()) {
            answer += plus.poll();
        }

        while (minus.size() > 1) {
            int first = minus.poll();
            int second = minus.poll();

            answer += first * second;
        }

        if (!minus.isEmpty()) {
            if (zero < 1) {
                answer += minus.poll();
            }
        }

        answer += one;

        System.out.println(answer);
    }
}
